torch/fx/experimental/symbolic_shapes.py

```diff
@@ -3069,40 +3073,62 @@ def issue_guard(guard: ShapeGuard) -> None:
-                # exclude sys.maxsize from our bounds to deal with direct
-                # == INT_MAX guards, but it's still dumb to actually test it.
-                # Note that you can be off by a pretty large constant and it
-                # won't matter because sizes in practice will be no where near
-                # the 64-bit limit.
-                if r.upper != sympy.oo and r.upper < sys.maxsize - 1 and g_upper is None:
-                    if any(is_dim(source) for source in sources):
-                        self.dim_constraints.add(sympy.Le(symbol, r.upper))
-                    bounds.append(str(r.upper))
-                if len(bounds) > 1:
-                    exprs.append(" <= ".join(bounds))
+            bounds.append(source_ref(sources[0]))
+            # NB: This looks like an off-by-one error but it's not: the
+            # upper bound may be sys.maxsize - 1 because we intentionally
+            # exclude sys.maxsize from our bounds to deal with direct
+            # == INT_MAX guards, but it's still dumb to actually test it.
+            # Note that you can be off by a pretty large constant and it
+            # won't matter because sizes in practice will be no where near
+            # the 64-bit limit.
+            if r.upper != sympy.oo and r.upper < sys.maxsize - 1 and g_upper is None:
+                if any(is_dim(source) for source in sources):
+                    self.dim_constraints.add(sympy.Le(symbol, r.upper))
+                # nontrivial upper bound is always interesting
+                bounds.append(str(r.upper))
+            if len(bounds) > 1:
+                exprs.append(" <= ".join(bounds))
+
+                # Check constraints
+                constraints = symbol_to_constraints[symbol]
+                for c in constraints:
+                    if isinstance(c, StrictMinMaxConstraint):
+                        # NB: By default, we have a restrictive range
+                        # 2 <= s0 <= sys.maxsize - 1.  But export users generally
+                        # expect to be able to specify nice ranges like [0, oo]
+                        if not (c.vr & self._default_value_range()).issubset(r):
+                            source = sources[0]
+
+                            expr = sympy.And(sympy.Le(r.lower, symbol), sympy.Le(symbol, r.upper))
+                            guard_expr = ShapeGuardPrinter(symbol_to_source, source_ref, self.var_to_sources).doprint(expr)
+                            var_with_range = self._render_range_for_constraint_violation(source, c)
+                            msg = (
+                                f"Not all values of {var_with_range} satisfy the generated guard {guard_expr}"
+                            )
+                            record_constraint_violation(
+                                c.warn_only,
+                                self._debug_name(source),
+                                msg,
+                            )
```

Comment:
@avikchaudhuri new constraint checking code.

Reply:
lgtm! 

I'd like to rationalize `sympy.oo` vs. `sys.maxint - 1` (for export I think we encode upper bounds as the latter but even if we didn't, they become the latter when they go into `var_to_range`) so I don't think you'll see `sympy.oo` here when doing `c.vr & ...` for export errors; but good to have it anyway. Also FYI we haven't done `0` lower bound yet, @pianpwk will probably do it soon. 

Curious: what's the status of `specialize_zero_one`, does size-oblivious have any bearing on it? Asking because `_default_value_range` switches based on it.

Reply:
I actually had to add the `&` to fix an export test lol, so it's definitely needed (although I feel like the current `>= 2` constraint was the most laod bearing.

Re `specialize_zero_one`, this is kind of... very misleading. What this does is say if the online solver should *eagerly* specialize on 0/1 quantities (and make the negative inference when the hint is not these). In practice, you will pretty quickly end up with this specialization anyway because of Tensor constructor guards. But maybe sometimes you wouldn't, and toggling this off would help you be less specialized in that case.